File: src/main/go/leetcode/editor/cn/56/MergeIntervals.go
//56.merge-intervals
//以数组 intervals 表示若干个区间的集合，其中单个区间为 intervals[i] = [starti, endi] 。请你合并所有重叠的区间，并返
//回一个不重叠的区间数组，该数组需恰好覆盖输入中的所有区间。
//
//
//
// 示例 1：
//
//
//输入：intervals = [[1,3],[2,6],[8,10],[15,18]]
//输出：[[1,6],[8,10],[15,18]]
//解释：区间 [1,3] 和 [2,6] 重叠, 将它们合并为 [1,6].
//
//
// 示例 2：
//
//
//输入：intervals = [[1,4],[4,5]]
//输出：[[1,5]]
//解释：区间 [1,4] 和 [4,5] 可被视为重叠区间。
//
//
//
// 提示：
//
//
// 1 <= intervals.length <= 104
// intervals[i].length == 2
// 0 <= starti <= endi <= 104
//
// Related Topics 数组 排序
// 👍 997 👎 0

package main

import (
	"fmt"
	"sort"
)

// leetcode submit region begin(Prohibit modification and deletion)
func merge(intervals [][]int) [][]int {
	var compare = func(a []int, b []int) bool {
		return a[0] < b[0] || a[0] == b[0] && a[1] < b[1]
	}
	sort.Slice(intervals, func(i int, j int) bool {
		return compare(intervals[i], intervals[j])
	})
	var ans [][]int
	for i, c := range intervals {
		if i == 0 {
			ans = append(ans, c)
		} else {
			var last = ans[len(ans)-1]
			//重叠
			if last[1] >= c[0] {
				last[1] = max_56(last[1], c[1])
			} else {
				ans = append(ans, c)
			}
		}
	}
	return ans
}

func max_56(x int, y int) int {
	if x > y {
		return x
	}
	return y
}

//leetcode submit region end(Prohibit modification and deletion)

// test from here
func main() {
	var arr = [][]int{{1, 3}, {2, 6}, {8, 10}, {15, 18}}
	fmt.Println(merge(arr))
	arr = [][]int{{1, 4}, {5, 6}}
	fmt.Println(merge(arr))
}

File: src/main/java/leetcode/editor/en/MergeIntervals.java
//56.merge-intervals
/**
 * Given an array of intervals where intervals[i] = [starti, endi], merge all
 * overlapping intervals, and return an array of the non-overlapping intervals that
 * cover all the intervals in the input.
 * <p>
 * <p>
 * Example 1:
 * <p>
 * <p>
 * Input: intervals = [[1,3],[2,6],[8,10],[15,18]]
 * Output: [[1,6],[8,10],[15,18]]
 * Explanation: Since intervals [1,3] and [2,6] overlap, merge them into [1,6].
 * <p>
 * <p>
 * Example 2:
 * <p>
 * <p>
 * Input: intervals = [[1,4],[4,5]]
 * Output: [[1,5]]
 * Explanation: Intervals [1,4] and [4,5] are considered overlapping.
 * <p>
 * <p>
 * Example 3:
 * <p>
 * <p>
 * Input: intervals = [[4,7],[1,4]]
 * Output: [[1,7]]
 * Explanation: Intervals [1,4] and [4,7] are considered overlapping.
 * <p>
 * <p>
 * <p>
 * Constraints:
 * <p>
 * <p>
 * 1 <= intervals.length <= 10⁴
 * intervals[i].length == 2
 * 0 <= starti <= endi <= 10⁴
 * <p>
 * <p>
 * Related Topics Array Sorting 👍 24522 👎 902
 */

package leetcode.editor.en;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;

public class MergeIntervals {
    public static void main(String[] args) {
        Solution solution = new MergeIntervals().new Solution();
    }

    //leetcode submit region begin(Prohibit modification and deletion)
    class Solution {
        public int[][] merge(int[][] intervals) {
            Arrays.sort(intervals, Comparator.comparingInt(x -> x[0]));
            int start = intervals[0][0];
            int end = intervals[0][1];
            List<int[]> res = new ArrayList<>();
            for (int i = 1; i < intervals.length; i++) {
                int[] e = intervals[i];
                if (end >= e[0]) {
                    end = Math.max(end, e[1]);
                } else {
                    res.add(new int[]{start, end});
                    start = e[0];
                    end = e[1];
                }
            }
            res.add(new int[]{start, end});
            return res.toArray(new int[0][]);

        }
    }
//leetcode submit region end(Prohibit modification and deletion)

}